### ISSUE TITLE
[Bug]: Global Twig Variable "editmode" does not exist.

### ISSUE THREAD

**Issue by solverat** [pimcore/pimcore#10504]
### Expected behavior

Rendering partials should always be possible. 

### Actual behavior

If you're using the delegated templating engine before the `KernelEvents::CONTROLLER` state, additional renderings will fail (if variable `editmode` is given).

### Steps to reproduce

Stupid example but easy to reproduce: Use this class, create any object and try to save:

![image](https://user-images.githubusercontent.com/700119/136032368-7ac10a3c-f8c1-4c7d-be7d-0c58911254d1.png)

```php
<?php

namespace App\EventListener;

use Pimcore\Event\DataObjectEvents;
use Pimcore\Event\Model\DataObjectEvent;
use Pimcore\Mail;
use Pimcore\Model\Document\Email;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;
use Symfony\Component\Templating\EngineInterface;

class TestListener implements EventSubscriberInterface
{
    public EngineInterface $template;

    public function __construct(EngineInterface $template)
    {
        $this->template = $template;
    }
    
    public static function getSubscribedEvents()
    {
        return [
            DataObjectEvents::PRE_UPDATE => 'onUpdate',
        ];
    }

    public function onUpdate(DataObjectEvent $event)
    {
         $data = $this->template->render(
            'example/data.html.twig',
            ['data' => []]
        );

         // use a template with a containing editmode variable
         $doc = Email::getByPath('/path/to/email');

         $mail = new Mail();

         $mail->setParam('data', $data);
         $mail->setDocument($doc);
         $mail->send();
    }
}
```

This happens because of this line:

https://github.com/pimcore/pimcore/blob/4d10439eb75b76acb8c1e07007f57d0d4e43078f/bundles/CoreBundle/EventListener/Frontend/GlobalTemplateVariablesListener.php#L75 

=> `$this->extensionSet->isInitialized()` will be `true` after the first render task (`$data`) and therefor this exception will be thrown after the second render task (email template).

**Comment by solverat** [comment]
Of course, this also counts for the global variable "document".

**Comment by brusch** [comment]
Well, this variables are not available at this point .. so what to do? 
I'd say just not use them in this context ...

**Comment by solverat** [comment]
"not use them in this context": In my example, the exception occurs while rendering the email template, which is definitely allowed to use variables like `editmode` or `document`. You then should say "It is not possible to render partials before any controller dispatches" which is a difficult statement, but that's my personal opinion of course.

**Comment by brusch** [comment]
Well, PRs "fixing" that without any side-effects (see the reason why it's currently implemented as it is) are always welcome 😉 I just wanted to share that this is currently by design as it's not that easy and that we're not gonna work on that anytime soon.

**Comment by dpfaffenbauer** [comment]
maybe `KernelEvents::REQUEST` could resolve it? that is the first event triggered by symfony. controller event would be the next.

**Comment by brusch** [comment]
That's where it was before ... with other issues, e.g. see: #5905 and some other things ... can't find them right now. But like I said, not that easy and kind of a chicken or the egg problem. The current implementation is kind of a "least pain" solution, even if it has it's drawbacks. 

In this particular case it could help to check the status of the templating engine and do the emailing later on, e.g. on response or terminate.

**Comment by brusch** [comment]
A kind of a hacky way would be to add the global variables already with some default values on `KernelEvents::REQUEST` and then override them on controller - not nice, but would solve this particular issue.

**Comment by blankse** [comment]
You can use also a service as global variable: https://symfony.com/doc/current/templating/global_variables.html#referencing-services

So you can create an global Pimcore Object, but than the templates must call `pimcore.document`, `pimcore.editmode`. Maybe an idea for Pimcore 11 or additionally.

**Comment by brusch** [comment]
I'll quickly create a PR with the hacky way, maybe someone can test it in detail.

**Comment by brusch** [comment]
Here we go ... 
@solverat can you please check & confirm #10505 thanks!

**Comment by brusch** [comment]
@blankse service: yep, nice idea - would make things a lot easier and elegant.

**Comment by brusch** [comment]
Fixed by #10505
